pangeo_forge_recipes/transforms.py

```diff
@@ -443,37 +435,89 @@ class CombineReferences(beam.PTransform):
+        return MultiZarrToZarr(
+            references,
+            concat_dims=self.concat_dims,
+            identical_dims=self.identical_dims,
+            target_options=self.target_options,
+            remote_options=self.remote_options,
+            remote_protocol=self.remote_protocol,
+            **self.mzz_kwargs,
+        )
+
+    def handle_gribs(self, indexed_references: Tuple[Index, list[dict]]) -> Tuple[Index, dict]:
+        """Handles the special case of GRIB format files by combining multiple references."""
+
+        references = indexed_references[1]
+        idx = indexed_references[0]
+        if len(references) > 1:
+            ref = self.to_mzz(references).translate()
+            return (idx, ref)
+        elif len(references) == 1:
+            return (idx, references[0])
+        else:
+            raise ValueError("No references produced for {idx}. Expected at least 1.")
+
+    def bucket_by_position(
+        self, indexed_references: Tuple[Index, dict], global_position_min_max_count
+    ) -> Tuple[int, dict]:
+        """Assigns a bucket based on index position to help preserve data order."""
+        idx = indexed_references[0]
+        ref = indexed_references[1]
+        global_min, global_max, global_count = global_position_min_max_count
+        sort_dimension = self.concat_dims[-1]
+        position = idx.find_position(sort_dimension)
+        range_size = global_max - global_min
+        num_buckets = math.ceil(global_count / self.max_refs_per_merge)
+        bucket_size = range_size / num_buckets
+        bucket = int((position - global_min) / bucket_size)
+        return bucket, ref
+
+    def global_combine_refs(self, refs) -> fsspec.FSMap:
+        """Performs a global combination of references to produce the final dataset."""
+        return fsspec.filesystem(
+            "reference",
+            fo=self.to_mzz(refs).translate(),
+            storage_options={
+                "remote_protocol": self.remote_protocol,
+                "skip_instance_cache": True,
+            },
+        ).get_mapper()
+
+    def expand(self, reference_lists: beam.PCollection) -> beam.PCollection:
+        min_max_count_positions = (
+            reference_lists
+            | "Get just the positions"
+            >> beam.MapTuple(lambda k, v: k.find_position(self.concat_dims[-1]))
+            | "Get minimum/maximum positions" >> beam.CombineGlobally(MinMaxCountCombineFn())
+        )
+        return (
+            reference_lists
+            | "Handle special case of gribs" >> beam.Map(self.handle_gribs)
+            | "Bucket to preserve order"
+            >> beam.Map(
+                self.bucket_by_position,
+                global_position_min_max_count=beam.pvalue.AsSingleton(min_max_count_positions),
+            )
+            | "Group by buckets for ordering" >> beam.GroupByKey()
+            | "Distributed reduce" >> beam.MapTuple(lambda k, refs: self.to_mzz(refs).translate())
```

Comment:
I'm still curious if we know that it's faster to do this intermediate `to_mzz` call - if it is then I'm fine with leaving this part in but it seems to add complexity since we have to re-group by key.

Reply:
I don't have specific analysis in the beam context at hand but it appears to parallelize pretty well from what I've seen: https://github.com/fsspec/kerchunk/issues/200#issuecomment-1220547197
At this point the shuffle isn't so bad; mostly just passing around some json so I would expect the groupby to be pretty snappy. We could and perhaps should consider moving the groupbykey logic up to the point of Index/URL `PCollection` creation, where the shuffle would be extremely minimal but that may require a bit of thought in terms of what kind of flexibility we want there. At some point, I'll wager we will have to come up with a few alternative data localization strategies - perhaps even to get performant pyramiding?

Reply:
That video is 💯

Reply:
And damned relevant for building indexes of arbitrary precision (e.g. if you're trying to query data for processing 'zoom levels' on a pyramid)!
https://opensource.googleblog.com/2008/08/uzaygezen-multi-dimensional-indexing.html
https://link.springer.com/referenceworkentry/10.1007/978-0-387-39940-9_350
https://tildesites.bowdoin.edu/~ltoma/teaching/cs3225-GIS/fall17/Lectures/gis-spaceFC.pdf
https://www.grin.com/document/322063?lang=en
https://github.com/locationtech/sfcurve

And some prior art in roughly our space:
https://github.com/locationtech/geotrellis/blob/af71a4ae027bfeaa9a930bc086ca98186e701953/store/src/main/scala/geotrellis/store/index/hilbert/HilbertSpatialKeyIndex.scala#L49